moabb/datasets/base.py

```diff
@@ -654,3 +656,206 @@ def data_path(
+
+    The method ``_download_subject`` must be implemented in each subclass
+    (see its docstring for more details).
+
+    If necessary, the methods ``_get_path_search_params`` and
+    ``_get_read_extra_params`` can be implemented in the subclass.
+    """
+
+    def _get_path_search_params(self, subject: int | None) -> dict[str, Any]:
+        """Return the kwargs for the :func:`mne_bids.find_matching_paths` function."""
+        out = {
+            "extensions": [
+                ".con",
+                ".sqd",
+                ".pdf",
+                ".fif",
+                ".ds",
+                ".vhdr",
+                ".set",
+                ".edf",
+                ".bdf",
+                ".EDF",
+                ".snirf",
+                ".cdt",
+                ".mef",
+                ".nwb",
+            ],
+        }
+        if subject is not None:
+            out["subjects"] = str(subject)
+        return out
+
+    def _get_read_extra_params(
+        self,
+        subject: int,  # pylint: disable=unused-argument
+    ) -> dict[str, Any] | None:
+        """Return the ``extra_params`` argument for the :func:`mne_bids.read_raw_bids` function."""
+        return None
+
+    @staticmethod
+    def _find_matching_paths(root, **kwargs) -> list[mne_bids.BIDSPath]:
+        bids_paths = mne_bids.find_matching_paths(root=root, **kwargs)
+        # Remove JSON files manually (the ignore_json argument only arrives in mne-bids=0.16)
+        return [bids_path for bids_path in bids_paths if bids_path.extension != ".json"]
+
+    @abc.abstractmethod
+    def _download_subject(self, subject, path, force_update, update_path, verbose) -> str:
+        """Download the data of a single subject and return the local path to the ROOT of the BIDS dataset.
+
+        Returns
+        -------
+        root : str
+            Path to the ROOT of the BIDS dataset.
+        """
+        pass
+
+    def bids_paths(
+        self, subject, path=None, force_update=False, update_path=None, verbose=None
+    ) -> list[mne_bids.BIDSPath]:
+        root = self._download_subject(subject, path, force_update, update_path, verbose)
+        return self._find_matching_paths(
+            root=root, **self._get_path_search_params(subject)
+        )
+
+    def data_path(
+        self, subject, path=None, force_update=False, update_path=None, verbose=None
+    ):
+        bids_paths = self.bids_paths(subject, path, force_update, update_path, verbose)
+        return [bids_path.fpath for bids_path in bids_paths]
+
+    def _get_single_subject_data(self, subject):
+        bids_paths = self.bids_paths(subject)
+        data = {}
+        for bids_path in bids_paths:
+            raw = mne_bids.read_raw_bids(
+                bids_path, extra_params=self._get_read_extra_params(subject)
+            )
+            # Data needs to be preloaded for the filtering step of paradigms
+            raw.load_data()
```

Comment:
```suggestion
            # Data needs to be preloaded for the filtering step of paradigms
            raw.load_data()
```

Okay for now, but we need to remove this step. Full lazy in the future, or only load at the filtering stage.

Reply:
Yes agreed, I think it would be good to include it to the filtering step. I followed the easy solution